egs/librispeech/asr/simple_v1/train.py

```diff
@@ -0,0 +1,264 @@
+                                             10000, 0)
+    tot_scores = -k2.get_tot_scores(target_graph, True, False).sum()
+    if training:
+        optimizer.zero_grad()
+        tot_scores.backward()
+        clip_grad_value_(model.parameters(), 5.0)
+        optimizer.step()
+
+    objf = tot_scores.detach().cpu()
+    total_objf = objf.item()
+    total_frames = nnet_output.shape[0]
+
+    return total_objf, total_frames
+
+
+def get_validation_objf(dataloader, model, device, graph, symbols):
+    total_objf = 0.
+    total_frames = 0.  # for display only
+
+    model.eval()
+
+    for batch_idx, batch in enumerate(dataloader):
+        objf, frames = get_objf(batch, model, device, graph, symbols, False)
+        total_objf += objf
+        total_frames += frames
+
+    return total_objf, total_frames
+
+
+def train_one_epoch(dataloader, valid_dataloader, model, device, graph,
+                    symbols, optimizer, current_epoch, num_epochs):
+    total_objf = 0.
+    total_frames = 0.
+
+    model.train()
+    for batch_idx, batch in enumerate(dataloader):
+        curr_batch_objf, curr_batch_frames = get_objf(batch, model, device,
+                                                      graph, symbols, True,
+                                                      optimizer)
+
+        total_objf += curr_batch_objf
+        total_frames += curr_batch_frames
+
+        if batch_idx % 100 == 0:
+            logging.info(
+                'processing batch {}, current epoch is {}/{} '
+                'global average objf: {:.6f} over {} '
+                'frames, current batch average objf: {:.6f} over {} frames'.
+                format(
+                    batch_idx,
+                    current_epoch,
+                    num_epochs,
+                    total_objf / total_frames,
+                    total_frames,
+                    curr_batch_objf / curr_batch_frames,
+                    curr_batch_frames,
+                ))
+
+        if valid_dataloader and batch_idx % 1000 == 0:
+            total_valid_objf, total_valid_frames = get_validation_objf(
+                dataloader=valid_dataloader,
+                model=model,
+                device=device,
+                graph=graph,
+                symbols=symbols)
+            model.train()
+            logging.info(
+                'Validation average objf: {:.6f} over {} frames'.format(
+                    total_valid_objf / total_valid_frames, total_valid_frames))
+    return total_objf
+
+
+def main():
+    # load L, G, symbol_table
+    lang_dir = 'data/lang_nosp'
+    with open(lang_dir + '/L.fst.txt') as f:
+        L = k2.Fsa.from_openfst(f.read(), acceptor=False)
+
+    with open(lang_dir + '/G.fsa.txt') as f:
+        G = k2.Fsa.from_openfst(f.read(), acceptor=True)
+
+    with open(lang_dir + '/words.txt') as f:
+        symbol_table = k2.SymbolTable.from_str(f.read())
+
+    L = k2.arc_sort(L.invert_())
```

Comment:
I assume in L.fst.txt  ilabels are phone ids and olabels are word ids. If it is the case, why L needs to be inverted before composing with G?

Reply:
What we do below is `Intersect` instead of `Compose`, so they should have the same set of ilables.  
BTW,  in `k2`, generally we we don't have a real `FST` struct, instead, we only have `FSA` struct (i.e. every arc is [src_state, dest_state, lable, weight]). `olabels` and any other quantities which can be attached on `arc` will go to `aux_labels`, you can think `aux_labesl` as a vector of struct whose size is arc-num. For now, we only store `olabels` in `aux_labels`, but this design is more flexible as we can attach any quantity on arcs, not just `olabels`.

Reply:
@qindazhu I am unable to find a Python wrapper for `from_file()` in `k2/python/k2/fsa.py`.

Reply:
you can just call
```
 with open(`fsa.txt`) as f:
        Fsa = k2.Fsa.from_str(f.read(), acceptor=True)
```
We don't have a version `from_file` for Fsa, only for symbol_table.

Reply:
@qindazhu if `k2.intersect(A, B)` where A and B both have `aux_labels`. What will the `aux_labels ` field be in the resulting fsa? If I want to do A compose B, what should I do?

Reply:
@freewym currently that is not supported, it will crash.  But you can fix it by moving the aux_labels of B to a different name.  Currently the result would end up with 3 sets of labels in that case, the labels, aux_labels and new_labels.  If you really only wanted 2 of those you could, say, copy the new_labels to `labels` and delete `new_labels`.  Assuming that was the one you wanted as the 'labels'.

Reply:
...  if needed, later we can change the python-level wrapper of intersect to do that automatically, i.e. to emulate the normal behavior of compose for FSTs which discards the interior symbols that are matched.   But that might end up being a bit un-intuitive; I think our current behavior is actually more intuitive than the standard FST approach.

Reply:
I implemented `compose` in https://github.com/k2-fsa/k2/pull/282 but it was later removed.
The code is available at https://github.com/k2-fsa/k2/commit/99d7b14c363655f3a86d4a207b8180e9178f15ce#diff-e02b2cfd8da9130757f0c5a4c623125f76185c37680c67919999f65b2d450187R167

Reply:
> @freewym currently that is not supported, it will crash. But you can fix it by moving the aux_labels of B to a different name. Currently the result would end up with 3 sets of labels in that case, the labels, aux_labels and new_labels. If you really only wanted 2 of those you could, say, copy the new_labels to `labels` and delete `new_labels`. Assuming that was the one you wanted as the 'labels'.

@danpovey Assuming `C=k2.intersect(A, B)`, it looks like the renamed `aux_labels` in B (renamed as `temp_labels`)  does not appear in C (i.e., `hasattr(C, "temp_labels") == False`)